gson-problem/src/main/java/org/zalando/problem/ProblemAdapterFactory.java

```diff
@@ -0,0 +1,138 @@
+package org.zalando.problem;
+
+import com.google.gson.*;
+import com.google.gson.internal.Streams;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.*;
+
+/**
+ * Problem {@link TypeAdapterFactory}.
+ */
+public class ProblemAdapterFactory implements TypeAdapterFactory {
+
+    private static final Set<Class<? extends Problem>> BASE;
+
+    static {
+        BASE = new HashSet<>();
+        BASE.add(Problem.class);
+        BASE.add(Exceptional.class);
+        BASE.add(ThrowableProblem.class);
+        BASE.add(AbstractThrowableProblem.class);
+        BASE.add(DefaultProblem.class);
+    }
+
+    private final boolean stacktrace;
+    private final Map<URI, Class<? extends Problem>> subtypes;
+    private final StatusTypeAdapter statusAdapter;
+
+    public ProblemAdapterFactory() {
+        this(false);
+    }
+
+    public ProblemAdapterFactory(boolean stacktrace) {
+        this(stacktrace, Status.class);
+    }
+
+    @SafeVarargs
+    public <E extends Enum & StatusType> ProblemAdapterFactory(final Class<? extends E>... statusTypes) {
+        this(false, statusTypes);
+    }
+
+    @SafeVarargs
+    public <E extends Enum & StatusType> ProblemAdapterFactory(boolean stacktrace, final Class<? extends E>... statusTypes) {
+        this(stacktrace, buildIndex(statusTypes));
+    }
+
+    public ProblemAdapterFactory(boolean stacktrace, Map<Integer, StatusType> statuses) {
+        this.stacktrace = stacktrace;
+        this.subtypes = new HashMap<>(2);
+        statusAdapter = new StatusTypeAdapter(statuses);
+    }
+
+    @SafeVarargs
+    private static <E extends Enum & StatusType> Map<Integer, StatusType> buildIndex(
+            final Class<? extends E>... types) {
+        final Map<Integer, StatusType> index = new HashMap<>();
+
+        for (final Class<? extends E> type : types) {
+            for (final E status : type.getEnumConstants()) {
+                if (index.containsKey(status.getStatusCode())) {
+                    throw new IllegalArgumentException("Duplicate status codes are not allowed");
+                }
+                index.put(status.getStatusCode(), status);
+            }
+        }
+
+        return Collections.unmodifiableMap(index);
+    }
+
+    public <T extends Problem> ProblemAdapterFactory registerSubtype(Class<T> clazz, URI type) {
+        Objects.requireNonNull(clazz);
+        Objects.requireNonNull(type);
+        if (subtypes.containsKey(type) || BASE.contains(clazz)) {
+            throw new IllegalArgumentException("class & type must be unique");
+        }
+
+        subtypes.put(type, clazz);
+        return this;
```

Comment:
I'd prefer that `ProblemAdapterFactory` is immutable and returns a new instance here. Can be accompanied by a `@CheckReturnValue` annotation to highlight to users that the return value has to be used.